### PR TITLE
feat(kill session): adapt global client session and using block in oracle model in kill session

### LINKED ISSUE #1375: [Bug]: kill session result into 'not owner of the threadid' error while connect to ob cloud instance
### ODC version

all 

### OB version

all

### What happened?

kill session result into 'not owner of the threadid' error while connect to ob cloud instance

### What did you expect to happen?

kill session success

### Anything else we need to know?

_No response_

### Cloud

_No response_

### PR DESCRIPTION
**Pull request by zijiacj**
#### What type of PR is this?
type-feature

#### Which issue(s) this PR fixes:

Fixes #1375

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 1]
```diff
@@ -211,6 +216,69 @@ public List<JdbcGeneralResult> executeKillSession(ConnectionSession connectionSe
...
             } catch (Exception e) {
                 if (StringUtils.contains(e.getMessage(), "Unknown thread id")) {
+                    // use 'select proxy_version()' command to check whether ob_proxy exists
+                    String obProxyVersion;
+                    try {
+                        obProxyVersion =
+                                connectionSession.getSyncJdbcExecutor(ConnectionSessionConstants.BACKEND_DS_KEY)
+                                        .queryForObject("select proxy_version()", String.class);
+                    } catch (Exception e1) {
+                        if (StringUtils.contains(e1.getMessage(), "FUNCTION proxy_version does not exist")) {
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:226]: I will first find out whether there is a view version command suitable for all odp versions, if there is no here, do not do early end, change to the same as before.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/KillSessionOBProxyVersionNumber.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * {@link }
+ *
+ * @description:
+ * @author: zijia.cj
+ * @date: 2024/7/19
+ * @since: ODC_release_4.3.1
+ */
+public enum KillSessionOBProxyVersionNumber {
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/KillSessionOBProxyVersionNumber.java:26]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 3]
```diff
@@ -211,6 +218,69 @@ public List<JdbcGeneralResult> executeKillSession(ConnectionSession connectionSe
...
+                                                        for (int i = 1; i <= columnCount; i++) {
+                                                            if ("TIME".equalsIgnoreCase(metaData.getColumnName(i))) {
+                                                                return true;
+                                                            }
+                                                        }
+                                                        return false;
+                                                    });
+                            if (enabledGlobalClientSession) {
+                                log.info(
+                                        "The obProxy has started the global client session, return error result directly");
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:258]: not 'started', but global client session enabled

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 4]
```diff
@@ -78,6 +81,10 @@ public class DefaultDBSessionManage implements DBSessionManageFacade {
             + "(?<port>[0-9]{1,5})\\*/.*";
     private static final Pattern SERVER_PATTERN = Pattern.compile(SERVER_REGEX);
     private static final ConnectionMapper CONNECTION_MAPPER = ConnectionMapper.INSTANCE;
+    private static final String GLOBAL_CLIENT_SESSION_OB_PROXY_VERSION_NUMBER = "4.2.3";
+    private static final String GLOBAL_CLIENT_SESSION_OB_VERSION_NUMBER = "4.2.2";
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:85]: image

[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:85]: The version recorded in the documentation is OB v4.2.2

[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:85]: is OB v4.2.5

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 5]
```diff
@@ -198,33 +210,182 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+        return finalResults;
+    }
+
+    private static boolean checkObServerDirected(ConnectionSession connectionSession) {
+        String sql;
+        if (OB_MYSQL == connectionSession.getDialectType()) {
+
+            sql = "select PROXY_SESSID from oceanbase.gv$ob_processlist where ID =(select id from oceanbase.gv$ob_processlist where info like '%gv$ob_processlist%');";
+        } else {
+            sql = "select PROXY_SESSID from gv$ob_processlist where ID =(select id from gv$ob_processlist where info like '%gv$ob_processlist%');";
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:270]: both 2 query for mysql mode and oracle mode may got error result.

there may exits multiple connection link, e.g. one via directly connect, one via obproxy,
current condition may match more than one result, then queryForObject will throw exception.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 6]
```diff
@@ -198,33 +210,182 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+                    .execute(anonymousCodeBlock);
+            return JdbcGeneralResult.successResult(sqlTuple);
+
+        } catch (Exception e) {
+            log.warn("Failed to kill session by using anonymous code blocks: {}", e.getMessage());
+            return jdbcGeneralResult;
+        }
+    }
+
+    private JdbcGeneralResult tryDirectConnectObserver(ConnectionSession connectionSession,
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:379]: method naming: tryKillSessionViaDirectConnectObServer

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 7]
```diff
@@ -198,33 +210,182 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+
+    private String getObProxyVersion(ConnectionSession connectionSession, boolean isDirectedOBServer) {
+        if (isDirectedOBServer) {
+            return null;
+        }
+        try {
+            return connectionSession.getSyncJdbcExecutor(ConnectionSessionConstants.BACKEND_DS_KEY)
+                    .queryForObject("select proxy_version()", String.class);
+        } catch (Exception e) {
+            log.warn("Failed to obtain the OBProxy version number: {}", e.getMessage());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:284]: log.info , here failed is expected for old version obproxy

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 8]
```diff
@@ -198,33 +210,182 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+            finalResults.add(jdbcGeneralResult);
+        }
+        return finalResults;
+    }
+
+    private static boolean checkObServerDirected(ConnectionSession connectionSession) {
+        String sql;
+        if (OB_MYSQL == connectionSession.getDialectType()) {
+
+            sql = "select PROXY_SESSID from oceanbase.gv$ob_processlist where ID =(select id from oceanbase.gv$ob_processlist where info like '%gv$ob_processlist%');";
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:268]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java (hunk 6 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:379]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java (hunk 5 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:270]: The result set will have only one record that determines whether the current session has an odp proxy on the link. so use queryForObject is ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java (hunk 7 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:284]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java (hunk 5 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:270]: But to get the original result set, and then make a unique judgment on the number of records, this is more robust, I will do that.

**Review by yizhouxw — DISMISSED** [review]
lgtm

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 9]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+     * @param results
+     * @return
+     */
+    private List<JdbcGeneralResult> processResults(ConnectionSession connectionSession,
+            List<JdbcGeneralResult> results) {
+        Boolean isDirectedOBServer = checkObServerDirected(connectionSession);
+        String obProxyVersion = getObProxyVersion(connectionSession, isDirectedOBServer);
+        String obVersion = (String) connectionSession.getAttribute("OB_VERSION");
+        boolean isEnabledGlobalClientSession =
+                checkGlobalClientSessionEnabled(connectionSession, obProxyVersion, obVersion);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:243]: isGlobalClientSessionEnabled

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 10]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+        boolean isSupportedOracleModeKillSession = checkOracleModeKillSessionSupported(obVersion, connectionSession);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:244]: isOracleModeKillSessionSupported

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 11]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+     * process the execution result after the first kill commands. if the result contains unknown thread
+     * id exception, try to use other solutions to execute the kill commands.
+     * 
+     * @param connectionSession
+     * @param results
+     * @return
+     */
+    private List<JdbcGeneralResult> processResults(ConnectionSession connectionSession,
+            List<JdbcGeneralResult> results) {
+        Boolean isDirectedOBServer = checkObServerDirected(connectionSession);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:239]: we should consider ob and non-ob, the logic below should be only for ob.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 12]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+                    .query("show proxyconfig like '" + configName + "';",
+                            rs -> rs.next() ? rs.getInt("value") : null);
+        } catch (Exception e) {
+            log.warn("Failed to obtain the value of OBProxy's configuration variable: {}", e.getMessage());
+            return null;
+        }
+    }
+
+    private boolean isUnknownThreadIdError(Exception e) {
+        return StringUtils.contains(e.getMessage(), "Unknown thread id");
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:353]: containsIgnorseCase

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 13]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+        }
+        if (isSupportedOracleModeKillSession) {
+            return tryKillSessionWithAnonymousBlock(connectionSession, jdbcGeneralResult,
+                    jdbcGeneralResult.getSqlTuple());
+        }
+        return tryKillSessionViaDirectConnectObServer(connectionSession, jdbcGeneralResult,
+                jdbcGeneralResult.getSqlTuple());
+    }
+
+    private boolean checkOracleModeKillSessionSupported(String obVersion, ConnectionSession connectionSession) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:375]: isOracleModeKillSessionSupported

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 14]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+    /**
+     * Try to kill session by using anonymous code blocks. If successful, return a successful
+     * jdbcGeneralResult, otherwise return the original jdbcGeneralResult.
+     *
+     * @param connectionSession
+     * @param jdbcGeneralResult
+     * @param sqlTuple
+     * @return
+     */
+    private JdbcGeneralResult tryKillSessionWithAnonymousBlock(ConnectionSession connectionSession,
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:391]: tryKillSessionByAnonymousBlock

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 15]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
+     * 
+     * @param connectionSession
+     * @param results
+     * @return
+     */
+    private List<JdbcGeneralResult> processResults(ConnectionSession connectionSession,
+            List<JdbcGeneralResult> results) {
+        Boolean isDirectedOBServer = checkObServerDirected(connectionSession);
+        String obProxyVersion = getObProxyVersion(connectionSession, isDirectedOBServer);
+        String obVersion = (String) connectionSession.getAttribute("OB_VERSION");
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:241]: ok

**Comment by yhilmare** [comment]
great job, the logic is very clear

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java [hunk 16]
```diff
@@ -198,33 +210,226 @@ private List<KillSessionResult> doKillSessionOrQuery(
...
-                if (StringUtils.contains(e.getMessage(), "Unknown thread id")) {
-                    try {
-                        log.info("Kill query/session Unknown thread id error, try direct connect observer");
-                        directLinkServerAndExecute(sqlTuple.getExecutedSql(),
-                                connectionSession);
-                        return JdbcGeneralResult.successResult(sqlTuple);
-                    } catch (Exception e1) {
-                        log.warn("Failed to direct connect observer {}", e1.getMessage());
-                    }
+                if (isUnknownThreadIdError(e)) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/session/DefaultDBSessionManage.java:250]: why shouldn't we judge the exception first and check these info secondly?

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**